server_test.go

```diff
@@ -3189,77 +3166,182 @@ func TestServerSubscribe(t *testing.T) {
 	tt := []struct {
 		desc    string
 		filter  string
-		handler func(topic string, pk packets.Packet)
-		wasNew  bool
+		client  string
+		handler func(client string, pk packets.Packet)
+		expert  error
 	}{
 		{
 			desc:    "subscribe",
 			filter:  "a/b/c",
+			client:  "client-1",
 			handler: handler,
-			wasNew:  true,
+			expert:  nil,
 		},
 		{
 			desc:    "subscribe existed",
 			filter:  "a/b/c",
+			client:  "client-1",
 			handler: handler,
-			wasNew:  false,
+			expert:  nil,
+		},
+		{
+			desc:    "subscribe different client id",
+			filter:  "a/b/c",
+			client:  "client-2",
+			handler: handler,
+			expert:  nil,
 		},
 		{
 			desc:    "subscribe case sensitive didnt exist",
 			filter:  "A/B/c",
+			client:  "client-1",
 			handler: handler,
-			wasNew:  true,
+			expert:  nil,
 		},
 		{
 			desc:    "wildcard+ sub",
 			filter:  "d/+",
+			client:  "client-1",
 			handler: handler,
-			wasNew:  true,
+			expert:  nil,
 		},
 		{
 			desc:    "wildcard# sub",
 			filter:  "d/e/#",
+			client:  "client-1",
+			handler: handler,
+			expert:  nil,
+		},
+		{
+			desc:    "subscribe $SYS/info",
+			filter:  "$SYS/info",
+			client:  "client-1",
+			handler: handler,
+			expert:  nil,
+		},
+		{
+			desc:    "subscribe invalied ###",
+			filter:  "###",
+			client:  "client-1",
+			handler: handler,
+			expert:  packets.ErrTopicFilterInvalid,
+		},
+		{
+			desc:    "subscribe invalied " + SharePrefix + "/#",
+			filter:  SharePrefix + "/#",
+			client:  "client-1",
 			handler: handler,
-			wasNew:  true,
+			expert:  packets.ErrTopicFilterInvalid,
+		},
+		{
+			desc:    "invalid handler",
+			filter:  "a/b/c",
+			client:  "client-1",
+			handler: nil,
+			expert:  packets.ErrInlineSubscriptionHandlerInvalid,
+		},
+		{
+			desc:    "invalid client id",
+			filter:  "a/b/c",
+			client:  "",
+			handler: handler,
+			expert:  packets.ErrClientIdentifierNotValid,
 		},
 	}
 
 	for _, tx := range tt {
 		t.Run(tx.desc, func(t *testing.T) {
-			require.Equal(t, tx.wasNew, s.Subscribe(tx.filter, tx.handler))
+			require.Equal(t, tx.expert, s.Subscribe(tx.client, tx.filter, tx.handler))
 		})
 	}
 }
 
 func TestServerUnsubscribe(t *testing.T) {
-	handler := func(topic string, pk packets.Packet) {
+	handler := func(client string, pk packets.Packet) {
 		// handler logic
 	}
 
+	client := "client-1"
```

Comment:
Probably should be `clientID`

```suggestion
	clientID := "client-1"
```